File: backend/definition/catalog/field_schema.py
from __future__ import annotations

from typing import Any

VALID_FIELD_TYPES = frozenset(
    {"string", "template", "text", "integer", "boolean", "enum", "json", "string_array"}
)


def _field_id(field: dict[str, Any]) -> str:
    return str(field.get("id") or "")


def coerce_field_value(field: dict[str, Any], raw: Any) -> Any:
    field_type = str(field.get("type") or "string")
    if raw is None:
        return None
    if field_type in {"string", "template", "text"}:
        return str(raw)
    if field_type == "integer":
        if isinstance(raw, bool):
            raise ValueError("boolean is not integer")
        if isinstance(raw, int):
            return raw
        text = str(raw).strip()
        if not text:
            return None
        return int(text)
    if field_type == "boolean":
        if isinstance(raw, bool):
            return raw
        if isinstance(raw, str):
            lowered = raw.strip().lower()
            if lowered in ("true", "1", "yes"):
                return True
            if lowered in ("false", "0", "no", ""):
                return False
        return bool(raw)
    if field_type == "enum":
        return str(raw)
    if field_type == "json":
        if isinstance(raw, (dict, list)):
            return raw
        if isinstance(raw, str):
            import json

            return json.loads(raw)
        return raw
    if field_type == "string_array":
        if isinstance(raw, list):
            return [str(item) for item in raw]
        if isinstance(raw, str):
            import json

            parsed = json.loads(raw)
            if isinstance(parsed, list):
                return [str(item) for item in parsed]
        raise ValueError("expected string array")
    return raw


def apply_defaults(schema: list[dict[str, Any]], data: dict[str, Any] | None) -> dict[str, Any]:
    merged: dict[str, Any] = {}
    for field in schema:
        field_id = _field_id(field)
        if not field_id:
            continue
        if "default" in field:
            merged[field_id] = field["default"]
    if data:
        merged.update(data)
    return merged


def validate_fields(
    schema: list[dict[str, Any]],
    data: dict[str, Any] | None,
    *,
    prefix: str = "",
    unknown_as_warning: bool = True,
) -> tuple[list[str], list[str]]:
    errors: list[str] = []
    warnings: list[str] = []
    payload = data or {}
    known_ids = {_field_id(field) for field in schema if _field_id(field)}

    for field in schema:
        field_id = _field_id(field)
        if not field_id:
            continue
        label = f"{prefix}{field_id}" if prefix else field_id
        value = payload.get(field_id)
        if field.get("required") and (value is None or value == ""):
            errors.append(f"{label} is required")
            continue
        if value is None:
            continue
        field_type = str(field.get("type") or "string")
        if field_type not in VALID_FIELD_TYPES:
            warnings.append(f"{label}: unknown field type '{field_type}'")
            continue
        try:
            coerced = coerce_field_value(field, value)
        except (TypeError, ValueError) as exc:
            errors.append(f"{label}: {exc}")
            continue
        if field_type == "enum":
            options = field.get("enum") or []
            if coerced not in options:
                errors.append(f"{label} must be one of {options}")
        if field_type == "integer" and not isinstance(coerced, int):
            errors.append(f"{label} must be integer")
        if field_type == "boolean" and not isinstance(coerced, bool):
            errors.append(f"{label} must be boolean")

    for key in payload:
        if key not in known_ids:
            msg = f"{prefix}{key}: unknown field"
            if unknown_as_warning:
                warnings.append(msg)
            else:
                errors.append(msg)

    return errors, warnings

File: frontend/src/shared/schema/fieldSchema.ts
import type { ToolFieldDef, ToolFieldType } from "@/features/tools/types/tool";

export const VALID_FIELD_TYPES: ToolFieldType[] = [
  "string",
  "template",
  "text",
  "integer",
  "boolean",
  "enum",
  "json",
  "string_array",
];

export function coerceFieldValue(field: ToolFieldDef, raw: unknown): unknown {
  const fieldType = field.type ?? "string";
  if (raw == null) return null;
  if (fieldType === "string" || fieldType === "template" || fieldType === "text") {
    return String(raw);
  }
  if (fieldType === "integer") {
    if (typeof raw === "boolean") throw new Error("boolean is not integer");
    if (typeof raw === "number") return raw;
    const text = String(raw).trim();
    if (!text) return null;
    return Number(text);
  }
  if (fieldType === "boolean") {
    if (typeof raw === "boolean") return raw;
    const lowered = String(raw).trim().toLowerCase();
    if (lowered === "true" || lowered === "1" || lowered === "yes") return true;
    if (lowered === "false" || lowered === "0" || lowered === "") return false;
    return Boolean(raw);
  }
  if (fieldType === "enum") return String(raw);
  if (fieldType === "json") {
    if (typeof raw === "object") return raw;
    if (typeof raw === "string") return JSON.parse(raw);
    return raw;
  }
  if (fieldType === "string_array") {
    if (Array.isArray(raw)) return raw.map(String);
    if (typeof raw === "string") {
      const parsed = JSON.parse(raw);
      if (Array.isArray(parsed)) return parsed.map(String);
    }
    throw new Error("expected string array");
  }
  return raw;
}

export function applyDefaults(
  schema: ToolFieldDef[],
  data: Record<string, unknown> | undefined,
): Record<string, unknown> {
  const merged: Record<string, unknown> = {};
  for (const field of schema) {
    if (field.default !== undefined) merged[field.id] = field.default;
  }
  if (data) Object.assign(merged, data);
  return merged;
}

export function validateFields(
  schema: ToolFieldDef[],
  data: Record<string, unknown> | undefined,
  prefix = "",
): { errors: string[]; warnings: string[] } {
  const errors: string[] = [];
  const warnings: string[] = [];
  const payload = data ?? {};
  const known = new Set(schema.map((field) => field.id));

  for (const field of schema) {
    const label = prefix ? `${prefix}${field.id}` : field.id;
    const value = payload[field.id];
    if (field.required && (value === undefined || value === "")) {
      errors.push(`${label} is required`);
      continue;
    }
    if (value === undefined) continue;
    try {
      const coerced = coerceFieldValue(field, value);
      if (field.type === "enum" && field.enum && !field.enum.includes(String(coerced))) {
        errors.push(`${label} must be one of ${field.enum.join(", ")}`);
      }
      if (field.type === "integer" && typeof coerced !== "number") {
        errors.push(`${label} must be integer`);
      }
      if (field.type === "boolean" && typeof coerced !== "boolean") {
        errors.push(`${label} must be boolean`);
      }
    } catch (error) {
      errors.push(`${label}: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  for (const key of Object.keys(payload)) {
    if (!known.has(key)) warnings.push(`${prefix}${key}: unknown field`);
  }

  return { errors, warnings };
}

export function objectToFieldValues(
  schema: ToolFieldDef[],
  input: Record<string, unknown>,
): Record<string, string> {
  const values: Record<string, string> = {};
  for (const field of schema) {
    const raw = input[field.id];
    if (raw == null) {
      values[field.id] = "";
      continue;
    }
    if ((field.type === "json" || field.type === "string_array") && typeof raw === "object") {
      values[field.id] = JSON.stringify(raw, null, 2);
    } else {
      values[field.id] = String(raw);
    }
  }
  return values;
}

export function fieldValuesToObject(
  schema: ToolFieldDef[],
  values: Record<string, string>,
): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const field of schema) {
    const raw = values[field.id] ?? "";
    if (!raw.trim() && field.default !== undefined) {
      result[field.id] = field.default;
      continue;
    }
    if (!raw.trim()) continue;
    if (field.type === "json" || field.type === "string_array") {
      try {
        result[field.id] = JSON.parse(raw);
      } catch {
        result[field.id] = raw;
      }
    } else if (field.type === "integer") {
      result[field.id] = Number(raw);
    } else if (field.type === "boolean") {
      result[field.id] = coerceFieldValue(field, raw);
    } else {
      result[field.id] = raw;
    }
  }
  return result;
}

export function genericFieldsFromInput(input: Record<string, unknown>): ToolFieldDef[] {
  return Object.keys(input).map((key) => ({
    id: key,
    type: typeof input[key] === "object" ? "json" : "template",
    scope: "call",
    ui: { label: key },
  }));
}
